test/extensions/access_loggers/grpc/http_grpc_access_log_integration_test.cc

```diff
@@ -346,5 +346,98 @@ TEST_P(AccessLogIntegrationTest, GrpcLoggerSurvivesAfterReloadConfig) {
+    HttpIntegrationTest::createUpstreams();
+    FakeUpstreamConfig config{timeSystem()};
+    config.upstream_protocol_ = Http::CodecType::HTTP2;
+    config.enable_manual_listen_ = true;
+    fake_upstreams_.emplace_back(std::make_unique<FakeUpstream>(0, version_, config));
+  }
+
+  void initialize() override {
+    config_helper_.addConfigModifier([](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+      auto* accesslog_cluster = bootstrap.mutable_static_resources()->add_clusters();
+      accesslog_cluster->MergeFrom(bootstrap.static_resources().clusters()[0]);
+      accesslog_cluster->set_name("accesslog");
+      ConfigHelper::setHttp2(*accesslog_cluster);
+    });
+
+    config_helper_.addConfigModifier(
+        [this](
+            envoy::extensions::filters::network::http_connection_manager::v3::HttpConnectionManager&
+                hcm) {
+          auto* access_log = hcm.add_access_log();
+          access_log->set_name("grpc_accesslog");
+
+          envoy::extensions::access_loggers::grpc::v3::HttpGrpcAccessLogConfig config;
+          auto* common_config = config.mutable_common_config();
+          common_config->set_log_name("foo");
+          common_config->set_transport_api_version(envoy::config::core::v3::ApiVersion::V3);
+          setGrpcService(*common_config->mutable_grpc_service(), "accesslog",
+                         fake_upstreams_.back()->localAddress());
+          common_config->mutable_grpc_stream_retry_policy()->mutable_num_retries()->set_value(10);
+          common_config->mutable_grpc_stream_retry_policy()
+              ->mutable_retry_back_off()
+              ->mutable_base_interval()
+              ->set_seconds(1);
+          access_log->mutable_typed_config()->PackFrom(config);
+        });
+
+    HttpIntegrationTest::initialize();
+  }
+};
+
+INSTANTIATE_TEST_SUITE_P(IpVersionsCientType, AccessLogRetryIntegrationTest,
+                         GRPC_CLIENT_INTEGRATION_PARAMS,
+                         Grpc::GrpcClientIntegrationParamTest::protocolTestParamsToString);
+
+TEST_P(AccessLogRetryIntegrationTest, BasicFlowWithRetryPolicy) {
+  SKIP_IF_GRPC_CLIENT(Grpc::ClientType::GoogleGrpc);
+  config_helper_.setDefaultHostAndRoute("foo.com", "/found");
+  initialize();
+  BufferingStreamDecoderPtr response = IntegrationUtil::makeSingleRequest(
+      lookupPort("http"), "GET", "/notfound", "", downstream_protocol_, version_);
+  ASSERT_TRUE(response->complete());
+  EXPECT_EQ("404", response->headers().getStatusValue());
+  test_server_->waitForCounterEq("cluster.accesslog.upstream_cx_connect_fail", 1);
+  test_server_->waitForCounterEq("cluster.accesslog.upstream_rq_retry", 1);
+  // If the number of retries you configured is small, when you reach here on the Test Thread, the
+  // number of remaining retries may already be 0, which may cause a Flaky Test. Here we have set
+  // the retries to 10 and the interval to 1 second, which is enough to prevent such a problem.
```

Comment:
This still seems like it has a high likelihood of flaking. Did you try running this with runs_per_test, etc. to make sure it doesn't flake? Especially under ASAN/TSAN?

One way of approaching this would be to inject a custom retry predicate that blocks until you are ready. A potentially better way of handling this would be to just have the fake log server return a 503, etc. to force a retry which then you fully control and you don't need any of this manual listen stuff.

Reply:
I'm not sure if it makes sense to have Fake Upstream return 503 because the gRPC access log retry is only triggered by `connect-failure`.

Even if we introduce a mechanism to block retry until listening starts immediately after `rq_retry` becomes `1` in the test thread, Envoy may start retrying before the mechanism is called. So, I think it is necessary to order the second and subsequent retry processes not to schedule retries until the listen starts, as shown in the following:

```cpp
{
// When a retry is executed for the second or later time, if Fake Upstream is not listening, do not schedule it; check the listening status every 5 seconds and schedule it if it is listening.
test_server_->disableRetryScheduleOverLimitedRetryCount(2, std::chrono::seconds(5), [] {
  // `listen_started_` is not necessarily true when Fake Upstream is listening, so it would need to be more complicated logic.
  return listen_started_;
});
test_server_->waitForCounterEq("cluster.accesslog.upstream_cx_connect_fail", 1);
test_server_->waitForCounterEq("cluster.accesslog.upstream_rq_retry", 1);
}
```
However, this implementation is going to be a bit more complicated and may require tweaking the retry logic of Envoy. This is not desirable.

Additionally, I also ran a test with `--runs_per_test=1000`, `--config=clang-tsan/clang-asan`, and found the FAILURE was 0/1000. I think this configuration will hardly cause Flaky. (If Counter is set to Eq, some failures were observed, so it changed it to Ge, but if the interval and number of retries are large enough, this is not a problem.)

Reply:
> via a custom retry predicate that just blocks until you are ready (via starting listening, etc.).

This design causes a number of problems. The `RetryPolicy` in the access logger uses something under the core namespace. (The access logger's `RetryPolicy` is under the core namespace by design, it could not depend on the `RetryPolicy` under `route_components.proto`), so it cannot use `retry_options_predicates`.

Reply:
Doesn't this code all go through the router? Have you implemented fully custom retry code?

I'm sorry but this is bound to flake so we will need to find a different way to test this. I'm open to other thoughts.

Reply:
It is difficult to fix flaky here, and we do not want to change the API. Even if I could use `RetryOptionPredicate`, I don't think there is a Socket Option to prevent Flaky. The reason why flakiness may be caused here is the presence of Jitter in the Retry BackOff. (If there is no Jitter, we can set a sufficiently large value for the interval and it should be fine.) However, changing the BackOff strategy is not a good idea either. So, I think the only way here is to set `flaky=True` and allow flaky.

Reply:
In general, any test that can flake will flake, so we have to figure out a way to fix this. I'm sorry. There must be some way of making this test non flaky. Can you please brainstorm how to do that? Perhaps you could latch on a stat? Maybe there is some way to force a retry if a 503 or some other error is sent that doesn't involve a connection failure? Etc.

Reply:
Fortunately, replacing `core.RetryPolicy` with `route.RetryPolicy` seems to work fine, thanks to API compatibility. How about testing the retry in this way for now?
I don't think it's possible to simulate `connect-failure` with any other implementation than the current one, because once a connection has been established with FakeUpstream, it will not be `connect-failure`.

Reply:
I think you can test the connect failure cases with a unit/mock test. Just figure out some other type of retry (503, H2 reset, whatever) that we can test deterministically using an integration test.